### ISSUE TITLE
Standardize Drasil File Structure

### ISSUE THREAD

**Issue by Mornix** [JacquesCarette/Drasil#1211]
Currently Drasil packages have no uniform "path" for inclusion. Some packages do one method, some do another.

Paths:
* drasil-lang: Language.Drasil
* drasil-data: Data.Drasil
* drasil-code: Language.Drasil.{Code,CodeSpec, etc}
* drasil-docLang: Drasil.{DocLang, DocumentLanguage, etc}
* drasil-gen: Language.Drasil
* drasil-printers: Language.Drasil
* drasil-example: Drasil.{GlassBR, SSP, etc}

From this list, we can see that packages split from drasil-lang retain the Language.Drasil prefix,
drasil-data has a pattern which is the same as drasil-lang, and drasil-docLang (and drasil-example although not "used" by anything) has a unique prefix with `Drasil1 at the top level.

It would be nice to standardize of a path structure for consistency, predictability, and before too many sub packages exist.

**Comment by JacquesCarette** [comment]
So there are at least two things going on here.  First, there is a division into "packages" -- these are supposed to be grouping together like things.  So we have
- lang: the representation language for two levels (basic information and chunks)
- data: fundamental knowledge that is shareable amongst many pieces of software
- code: code generation
- docLang: document language routines
- gen: generation of all artifacts
- printers: document generation
- example: examples of Drasil use

There is also some places to put modules related to Drasil.  It is customary, in Haskell, to define things that are "languages" in `Language.X` and "data" for X in `Data.X`. Because Drasil is itself a rich language, it makes some sense to use it as  namespace too, and so the examples fit there.  GHC, for example, has a top-level `GHC` package.

That does leave drasil-code, drasil-docLang, drasil-gen and drasil-printers as not having a solid namespace design. So I guess I would be quite open to suggestions (perhaps after perusing Hackage?) of where to place some of these things.

**Comment by samm82** [comment]
For the sake of completeness, we also have:
- build: build system languages and renderers

**Comment by Mornix** [comment]
So I've thought about this for a bit and took a very very brief look at Hackage. I noticed `Language.ABC` as you mentioned. I did also see pretty printers being placed under `Text.ABC`. So I would say with that bit of information:
drasil-build → Text.Drasil.Build
drasil-code → Text.Drasil.Code
drasil-printers → Text.Drasil.(Printers | just the individual languages like TeX and HTML — Note: to properly do this, we would need to clear up `DocType` having `SRS` and `Website` and move that to a more appropriate package for the `SRS` component.

https://github.com/JacquesCarette/Drasil/blob/master/code/drasil-printers/Language/Drasil/Output/Formats.hs#L11

drasil-gen → Language.Drasil.Gen
drasil-docLang → Language.Drasil.DocLang
drasil-database → Language.Drasil.Database

drasil-theory → Data.Drasil.Theory (since this seems to be more data oriented)

**Comment by JacquesCarette** [comment]
I agree with all of these proposed renamings. Doing it should be done quite incrementally though.

From printers, Text.Drasil.Tex (etc) is probably a good way to do it. And indeed, shouldn't be done until 1) the current work on indentation is finished, and 2) `DocType` is fixed.

**Comment by balacij** [comment]
Personally, I agree with @JacquesCarette's first comment:
> Because Drasil is itself a rich language, it makes some sense to use it as namespace too, and so the examples fit there. GHC, for example, has a top-level GHC package.

And that is because of ~2 things:
1. The base-level `Data.X`, `Language.X`, etc. packages are usually modifiers of or define new (generic) data types that are usable in any project. Ours are really meant for use with Drasil, and would be awkward to use in other applications.
2. I like a separation between "what things are really core to Drasil and what things are needed of our host language".

For reason #2, I think the paths in `drasil-utils` should be something like:
```
Data.String.Extras
Data.List.Extras
Text.PrettyPrint.Extras
etc.
```

**Comment by JacquesCarette** [comment]
I agree with your suggestions regarding `drasil-utils`

**Comment by Purav-Kanda** [comment]
Taking a look at this. The current drasil-utils modules are:

Utils.Drasil.Lists — replaceAll, subsetOf, nubSort, weave, foldle, foldle1, toColumn, mkTable
Utils.Drasil.Strings — toPlainName, repUnd
Utils.Drasil.English — capitalize, stringList
Utils.Drasil — re-export of all three

38 files import these. Most use import Utils.Drasil (someFunction) through the re-export.
Proposed mapping:

Utils.Drasil.Lists → Data.List.Extras
Utils.Drasil.Strings → Data.String.Extras
Utils.Drasil.English → Data.String.English? (not sure — it wasn't in the original list)
Utils.Drasil (re-export) → remove, update importers to use specific submodules

Does this look right, especially for English?

**Comment by JacquesCarette** [comment]
`Utils.Drasil.English` won't make sense. Really, we should separate all the natural language stuff (now mostly in `drasil-lang`) into `drasil-natural-language`, and then `Utils.Drasil.English` could be redistributed to be there (potentially in multiple places).

Note that the `Utils.Drasil` re-export can be fixed independently of the other renames. And `Data.List.Extras`, `Data.String.Extras` would still find a "best home" in `drasil-utils` as far as I'm concerned!

**Comment by Purav-Kanda** [comment]
Got it — I'll leave English for now and focus on renaming Lists → Data.List.Extras and Strings → Data.String.Extras within drasil-utils. I'll update the re-export module's internal imports but leave its name for a separate PR.
